### ISSUE TITLE
Make the plugin more consistent with jQuery UI Datepicker

### ISSUE THREAD

**Issue by benjamin-albert** [KidSysco/jquery-ui-month-picker#16]
It makes sense that the month picker plugin will be used side by side with jQuery UI's Datepicker, so it makes sense that it will behave like it (at least from the user's perspective).

Here is a list of behaviors that cannot be changed without changing the plugin's source:
<ul>
<li>
Animation effect and speed (Datepicker's default is a fadein animation that takes 400ms).
</li>
<li>Button image and show behavior (looks like I'm not the first one to mention this #5).</li>
<li> this doesn't refers to the associated input field in all of the API's callbacks.</li>
<li>The month format is not as flexible as the <a href='https://api.jqueryui.com/datepicker/#option-dateFormat'>dateFormat</a> option.</li>
<li>Pressing enter should choose the current date and close the menu.</li>
</ul>

If anyone finds other inconsistencies please point them out here.

I've already implemented a `Duration` option which behaves like Datepicker's <a href='http://api.jqueryui.com/datepicker/#option-duration'>duration</a> option which allows the user to choose from 3 predefined speeds:
<ul>
<li>normal (the default) 400ms</li>
<li>fast 200ms</li>
<li>slow 600ms</li>
</ul>

Or you can specify the speed in milliseconds by passing in a number.

The `Duration` option's implementation is similar to Datepicker's one (it's based on <a href='https://github.com/jquery/jquery/blob/3dd3d1357d027a487559516fcdd9064cc792edab/src/effects.js#L668'>$.fx.speeds</a>).

I'm not sure yet if I will add support for jQuery UI's effect's in the show/hide animation options seeing as it might require a workaround in order to get it to work with jQuery UI position.

I will post more about this later.

The version I'm using at my company already has a `Format` option it's implementation is based on <a href='http://api.jqueryui.com/datepicker/#utility-formatDate'>$.datepicker.formatDate</a>, but like i've said in my first pool request it needs a little polishing before I can make it open source.

I am planning to have two branches the master branch which all commits to it will not change the existing default behavior but will offer (most) of the options to make it be more consistent with datepicker, and another branch (I think I'll call it datepicker-consistent) which by default will aim to be as consistent as possible with Datepicker.

**Comment by KidSysco** [comment]
This sounds like a fine plan. But what options do you see differing between your 2 branches? Just default settings? I can just wait to see what you come up with too, it should be interesting.

**Comment by benjamin-albert** [comment]
The datepicker-consistent branch will by default:
<ul>
<li>Use a fade in/out animation.</li>
<li>Show and hide's speed will default to 'normal' aka 400ms</li>
<li>Have this refers to the associated input field in all callbacks</li>
</ul>

I might come up with more <i>breaking-ish</i> changes as I go.
If I come up with more I'll update you.

**Comment by KidSysco** [comment]
OK, well I am just not sure this is worth splitting off into another branch. As long as no breaking changes are made. Based on this, all unit tests should still pass.

If you really have a breaking change, that effects on of the unit tests or some other major piece of functionality, then I think we are OK with just one branch for now.

What do you think?

**Comment by benjamin-albert** [comment]
If you don't have any problems changing the default options I'd prefer have one branch.

**Comment by KidSysco** [comment]
Agreed!

**Comment by benjamin-albert** [comment]
I've created a pull request for most of the features I've mention here see: #17.

But I still have some more features I need to add before I can use the opens source version in my company.

I will finish implementing the all of the features I've mentioned here and then I will open another issue discussing the other feature I want to add.

**Comment by KidSysco** [comment]
I am going to leave this issue open for now actually...

I just added a bunch of tests to ensure everything is still working as expected. This is all my fault as there were no tests to show this stuff, and the issue may have been around before your code was introduced too.

I am not sure what happened but the tests for the HTML 5 month Input type were lost, so I added them back in and I am noticing that if the html 5 month input type is used, then it can no longer be populated. Take a look at the updated fiddle...

http://jsfiddle.net/kidsysco/JeZap/

I tried the MonthFormat option of 'yy-mm' which according to the docs you sent should give us something that looks like, 2016-01, which is the format the w3c has accepted for the control.

I can see the code in there on line 259 for this but it looks to be just set to default in this case and cannot be over-ridden as a result. So I think changing line 259 from...

this.options.MonthFormat = this.MonthInputFormat;

to 

this.options.MonthFormat = 'yy-mm';

will fix it.

But I was not entirely sure and I don't want to mess you up. So I thought I would run it past you.

**Comment by KidSysco** [comment]
I added more tests for the position work you did too checking all sides of the window ensuring no overlapping. 

I also added a bunch of tests for GetSelectedDate() returning a date object rather than strings. I love that one! Good job there.

**Comment by benjamin-albert** [comment]
Your right about the MonthInputFormat I copied and pasted the format from the default options in mistake.

MonthInputFormat Is defined under the ParseMonth method changing it to 'yy-mm' will solve the problem.

**Comment by KidSysco** [comment]
ok good, do you want to handle this in your next pull request? That way there is no merge conflict.

**Comment by benjamin-albert** [comment]
I'm almost ready to open a pull request for the <a href='https://github.com/benjamin-albert/jquery-ui-month-picker/tree/2.5>2.5 branch</a> in my fork.

If you don't mind waiting a day or two I'd prefer make the change myself and avoid potential merge conflicts.

**Comment by KidSysco** [comment]
OK yea, throw in that one last change just discussed to fix the html 5 month type. I just tested it in fiddle and it works perfectly!

**Comment by benjamin-albert** [comment]
It's already included see the <a href='https://github.com/benjamin-albert/jquery-ui-month-picker/commit/99bde89272c5e059d192a21f213c282df0b0191f'>latest commit to 2.5</a>.

The reason that this slipped is that when I tested out creating a subclass that uses <a href='http://momentjs.com'>Moment.js</a> for parsing and formatting I've assigned a correct `MonthInputFormat`.

This was a very small change I made while writing documentation for this feature and realized this would be required, so once I sew the HTML 5 month input worked with the Moment.js subclass I committed the change and proceeded with writing documentation.

I guess this shows the importance of unit tests.
